### PR TITLE
[GPU][BUGFIX]: Fixed bug in SliceOp::calc_output_layout.

### PR DESCRIPTION
**Pull request by pkowalc1**
### Details:
 - Fixed bug in SliceOp::calc_output_layout, which was calculating wrong output shape due to use of deprecated tensor struct.
 - Fixed and refactored all SliceOp unittests - which were expecting wrong output data

### Tickets:
 - CVS-131491

### CODE REVIEW

File: src/plugins/intel_gpu/src/graph/slice.cpp [hunk 1]
```diff
@@ -26,11 +26,12 @@ inline std::vector<layout> slice_inst::calc_output_layouts(const slice_node&, co
     std::unordered_map<size_t, ov::Tensor> const_data;
     for (std::size_t i = 1; i < impl_param.input_layouts.size(); i++) {
         auto shape_len = shape_size(impl_param.input_layouts[i].get<ShapeType>().to_shape());
-        input_shapes.push_back({static_cast<ov::Dimension::value_type>(shape_len)});
+        const ov::PartialShape input_shape{static_cast<ov::Dimension::value_type>(shape_len)};
+        input_shapes.push_back(input_shape);
```
Comment (praasz) [inline src/plugins/intel_gpu/src/graph/slice.cpp:30]:
```suggestion
        input_shapes.emplace_back({static_cast<ov::Dimension::value_type>(shape_len)});
```
then in `make_tensor` use `input_shapes.back()`.

It will avoid create additional `PartialShape` which is copied to `input_shapes` vector then copied to `layout` and destroyed at end.

File: src/tests/functional/shared_test_classes/src/single_op/slice.cpp [hunk 2]
```diff
@@ -59,6 +59,11 @@ void Slice8LayerTest::SetUp() {
         slice = std::make_shared<ov::op::v8::Slice>(param, begin_node, end_node, stride_node);
     }
     function = std::make_shared<ov::Model>(slice->outputs(), ov::ParameterVector{param}, "Slice-8");
+
+    // [PK]: Slice op only moves memory, no calculation takes places, so
```
Comment (praasz) [inline src/tests/functional/shared_test_classes/src/single_op/slice.cpp:63]:
```suggestion
    // Slice op only moves memory, no calculation takes places, so
```

File: src/plugins/intel_gpu/src/graph/slice.cpp [hunk 3]
```diff
@@ -17,7 +17,7 @@ slice_inst::typed_primitive_inst(network& network, slice_node const& node)
 
 layout slice_inst::calc_output_layout(slice_node const& node, kernel_impl_params const& impl_param) {
     auto l = calc_output_layouts<ov::PartialShape>(node, impl_param)[0];
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/graph/slice.cpp:19]: Fixed.

File: src/tests/functional/shared_test_classes/src/single_op/slice.cpp (hunk 2 above)
Comment (pkowalc1) [reply src/tests/functional/shared_test_classes/src/single_op/slice.cpp:63]: Fixed.

File: src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+data_types ToDataType<int64_t>()  { return data_types::i64; }
+
+// Generates buffer of the length of shape_size(shape).
+template<typename T>
+std::vector<T> GenInput(const ov::PartialShape& shape) {
+    const size_t size = ov::shape_size<ov::Shape>(shape.get_shape());
+    std::vector<T> result;
+    for (size_t i = 0; i < size; i++)
+        result.push_back(static_cast<T>(i));
+    return result;
```
Comment (praasz) [inline src/plugins/intel_gpu/tests/unit/test_cases/slice_gpu_test.cpp:41]:
```suggestion
    const auto size = ov::shape_size(shape.get_shape());
    auto result = std::vector<T>(size);
    std::iota(result.begin(), result.end(), T{0});
    return result;
```
If T is fundamental type then should be simpler and result vector would allocated once.

File: src/plugins/intel_gpu/src/graph/slice.cpp (hunk 1 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/src/graph/slice.cpp:30]: It won't work - input_shapes is of type Shape, not PartialShape, and Shape does not have a proper ctor. Even if you could make it, then layout class in make_tensor accepts only PartialShape, not Shape, so you would have additional ctor called.

IMO the way it is right know it is more readable, which is more important than such optimizations, especially  in a code that likely you will never see in any profiling.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+            ::testing::ValuesIn(static_params),
+            ::testing::Values(ov::element::i64),
+            ::testing::Values(ov::test::utils::DEVICE_GPU)),
+        Slice8LayerTest::getTestCaseName);
+
+INSTANTIATE_TEST_SUITE_P(
+        smoke_CLDNN_F32, Slice8LayerTest,
+        ::testing::Combine(
+            ::testing::ValuesIn(static_params),
+            ::testing::Values(ov::element::f32),
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:66]: 1. Looks like these 3 instances can be shrunk if you replace `Values` to `ValuesIn({f32, i64, etc})`
2. Please add instance for f16 precision too

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+        Slice8SpecificParams{ {{{}, {{ 1, 5, 32, 20 }}}}, { 0, 1, 0, 0 }, { 1, 3, 32, 20 }, { 1, 1, 1, 1 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 2, 5, 32, 20 }}}}, { 0, 0, 10, 0 }, { 1, 3, 20, 20 }, { 1, 1, 1, 1 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 1, 5, 32, 32 }}}}, { 0, 0, 20, 20 }, { 1, 5, 25, 26 }, { 1, 1, 1, 2 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 2, 5, 32, 32 }}}}, { 0, 0, 0, 20 }, { 1, 2, 30, 30 }, { 1, 1, 2, 1 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 1, 5, 32, 20 }}}}, { 0, 0, 2, 10 }, { 1, 3, 32, 20 }, { 1, 1, 1, 1 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 2, 5, 32, 32 }}}}, { 0, 1, 0, 10 }, { 1, 5, 32, 30 }, { 1, 1, 1, 1 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 1, 5, 32, 20 }}}}, { 0, 1, 2, 10 }, { 1, 5, 32, 18 }, { 1, 1, 1, 2 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 2, 8, 32, 20 }}}}, { 0, 0, 2, 10 }, { 1, 8, 32, 18 }, { 1, 2, 1, 2 }, { 0, 1, 2, 3 } },
+        Slice8SpecificParams{ {{{}, {{ 2, 8, 32, 20 }}}}, { 0, -20, -15 }, { 2, -5, 3 }, { 1, 1, 1 }, { 0, 2, 1 } }
```
Comment (vladimir-paramuzov) [inline src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:51]: Please ensure that all of those test cases are meaningful and don't just duplicate each other in terms of tested functionality. Maybe some of those can be moved to nightly scope or removed at all. 

Note: test cases with `*smoke*` prefix are executed on pre-commit stage and we don't want precommit time to increase significantly (it's already too slow). So need to be careful when adding new tests into smoke scope

File: src/tests/functional/shared_test_classes/src/single_op/slice.cpp [hunk 7]
```diff
@@ -59,6 +59,11 @@ void Slice8LayerTest::SetUp() {
...
+    // any difference likely means a bug.
+    rel_threshold = 0;
+    abs_threshold = 0;
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/tests/functional/shared_test_classes/src/single_op/slice.cpp:66]: IMO it will be wrong. At this level we should only test the operator itself. If there are any implicit conversions - they should be disabled for test(how to do it?)- otherwise we might not catch the error in slice operator if values are close enough.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+INSTANTIATE_TEST_SUITE_P(
+        smoke_CLDNN_F32, Slice8LayerTest,
+        ::testing::Combine(
+            ::testing::ValuesIn(static_params),
+            ::testing::Values(ov::element::f32),
+            ::testing::Values(ov::test::utils::DEVICE_GPU)),
+        Slice8LayerTest::getTestCaseName);
+
+INSTANTIATE_TEST_SUITE_P(
+        smoke_CLDNN_I32, Slice8LayerTest,
```
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:71]: Fixed.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp (hunk 5 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:66]: Fixed.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp (hunk 6 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:51]: Basically, imo we should prepare such data only once(ideally even in stochastic way) - and run the operators on all supported devices with the same data. Of course, there should be also possibility to added specific data for corner cases, but the point is that it should be created only once for all platforms supported. Simply - refactoring of those tests is needed, which is out of the scope of this task.

To be honest, I have just follow a convention that such data(and file) is prepared for each device separately. I have asked other people and they confirmed  me that this is what they do: they just copy the data in many cases :) 
Originally, I just wanted to modify the functional tests on cpu to run it on GPU as well, but I was asserted that it is not the convention here. 

About duplicating - generally speaking, data tests prepared for cpu device are good enough already IMO - they already test all the edge cases and it is hard to not duplicate them in that sense - GPU has to pass them.  The only thing that comes to my mind is that those inputs shapes should be generated randomly, but it would require refactoring, which is out of the scope of this task.

In other words, tests needs refactoring, which is out of the scope of this task, so I just limited myself to add comparison tests for now.

File: src/tests/functional/shared_test_classes/src/single_op/slice.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (vladimir-paramuzov) [reply src/tests/functional/shared_test_classes/src/single_op/slice.cpp:66]: For operator only testing unit tests should be used. 
Functional single layers tests run e2e plugin pipeline for trivial IRs using public API. There's no way to disable transformations (such as precision conversion) in that case.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp (hunk 6 above)
Comment (vladimir-paramuzov) [reply src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:51]: > Basically, imo we should prepare such data only once(ideally even in stochastic way) - and run the operators on all supported devices with the same data. 

Such approach is used in conformance tests. For functional tests there are no strict rules on test scope or input parameters - that is controlled per plugin with some reuse of common things.

 > I have asked other people and they confirmed me that this is what they do: they just copy the data in many cases

Each device/plugin may have specific cases which may impact the scope. For instance, GPU plugin needs f16 precision tests while CPU - doesn't. So there's nothing wrong with customization of functional tests scope for each plugin w.r.t the actual implementation/supported features/etc. 

> About duplicating - generally speaking, data tests prepared for cpu device are good enough already IMO - they already test all the edge cases and it is hard to not duplicate them in that sense - GPU has to pass them. The only thing that comes to my mind is that those inputs shapes should be generated randomly, but it would require refactoring, which is out of the scope of this task.

The thing is that model compilation for GPU is much slower than for CPU, and we can't introduce hundreds or even thousands of tests cases for each operation as testing time growth too fast. So if we can split test cases and keep only critical things for precommit scope and remove / move to nightly the rest, then ideally it should be done. I'd suggest checking exec time for current instances, and if it fits into 1-2m limit, then OK, we can keep it as is.

File: src/tests/functional/shared_test_classes/src/single_op/slice.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/tests/functional/shared_test_classes/src/single_op/slice.cpp:66]: Ok, so will remove it completely, but just wanted to mention that default thresholds(which are calculated in runtime, depending on data values), are absurdly big - e.g. 1. for integer cases, e.g. ~0.2 for floating point, and to my best knowledge(asked people about it) - there is no easy way to set it per case.

[the remark this answers is not in the document]
Comment (pkowalc1) [reply src/tests/functional/shared_test_classes/src/single_op/slice.cpp:66]: > Functional single layers tests run e2e plugin pipeline for trivial IRs using public API. There's no way to disable transformations (such as precision conversion) in that case.
In that case such reference tests should be moved to unit tests IMO. Of course nothing stands in the way to do such comparison tests at two levels - but higher level should focus on integration itself( with transformations, etc ), not on the validating operator itself. But again - it is outside the scope of this task to do such refactoring, so I will just remove setting that threshold.

File: src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp (hunk 5 above)
Comment (pkowalc1) [reply src/plugins/intel_gpu/tests/functional/shared_tests_instances/single_layer_tests/slice.cpp:66]: Fixed.

**Comment by pkowalc1** [comment]
build_jenkins

**Review by praasz — APPROVED** [review]

**Review by vladimir-paramuzov — APPROVED** [review]
